webapp/src/components/code_review/code_review.tsx

```diff
@@ -0,0 +1,194 @@
+    line?: number;
+};
+
+const Avatar = (window as any).Components.Avatar;
+const imageURLForUser = (window as any).Components.imageURLForUser;
+
+export default class CodeReview extends React.PureComponent<Props, State> {
+
+    static MAX_FILE_SIZE = 500000;
+
+    static SupportedFiles: { [extension: string]: any } = {
+        diff: {name: 'Diff'},
+        patch: {name: 'Diff'},
+    }
+
+    static getDerivedStateFromProps(props: Props, state: State) {
+        if (props.fileInfo.id !== state.prevFileId) {
+            const usedLanguage = props.fileInfo.extension in CodeReview.SupportedFiles ? props.fileInfo.extension : undefined;
+
+            if (!usedLanguage || props.fileInfo.size > CodeReview.MAX_FILE_SIZE) {
+                return {
+                    code: '',
+                    lang: '',
+                    loading: false,
+                    success: false,
+                    prevFileId: props.fileInfo.id,
+                };
+            }
+
+            return {
+                code: '',
+                lang: usedLanguage,
+                loading: true,
+                prevFileId: props.fileInfo.id,
+            };
+        }
+        return null
+    }
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            code: '',
+            lang: '',
+            loading: true,
+            success: true,
+        };
+    }
+
+    componentDidMount() {
+        this.load();
+    }
+
+    load = async () => {
+        try {
+            const fileInfo = this.props.fileInfo;
+            const fileUrl = getFileUrl(fileInfo.id);
+            if (!this.state.lang || fileInfo.size > CodeReview.MAX_FILE_SIZE) {
+                return;
+            }
+            const codeResponse = await fetch(fileUrl);;
+            const code = await codeResponse.text();
+            this.handleReceivedCode(code);
+        } catch (e) {
+            this.handleReceivedError();
+        }
+    }
+
+    handleReceivedCode = (code: string) => {
+        this.setState({
+            code,
+            loading: false,
+            success: true,
+        });
+    }
+
+    handleReceivedError = () => {
+        this.setState({loading: false, success: false});
+    }
+
+    static supports(fileInfo: FileInfo, post: Post) {
+        const isSupportedFile = fileInfo.extension in CodeReview.SupportedFiles;
+        const isRootPost = post.root_id === "";
+        return isSupportedFile && isRootPost;
+    }
+
+    handleMouseMove(event: React.MouseEvent<HTMLDivElement>) {
+        const target = event.currentTarget;
+
+        for (let line of target.parentElement!.querySelectorAll('.CodeReview-LineOverlay')) {
+            let element = line as HTMLElement;
+            element.classList.remove('CodeReview-LineOverlay-focused');
+        }
+
+        if (target.classList.contains('CodeReview-LineOverlay')) {
+            target.classList.add('CodeReview-LineOverlay-focused');
+        }
+    }
```

Comment:
This code is a little confusing, mainly because we're creating an `onMouseMove` callback on each item in an array of components, while this callback is examining the the physical dom to get the parent element of this component. Since there is this coupling, there's code in three different spots that rely on this logic.

I'm wondering if the state of the focused line overlay can be defined in a more React-friendly  state variable instead of css class, to be more explicit in all spots that will share this logic. This will allow the code all around to be simpler.